File: feature/feature-template/src/main/java/org/gampiot/robok/feature/template/code/CodeTemplate.java
package org.gampiot.robokide.feature.template.code;

import android.os.Parcel;
import android.os.Parcelable;

import androidx.annotation.NonNull;

public class CodeTemplate implements Parcelable {

      public static final Parcelable.Creator<CodeTemplate> CREATOR =
          new Parcelable.Creator<CodeTemplate>() {
               @Override
               public CodeTemplate createFromParcel(Parcel parcel) {
                    return new CodeTemplate(parcel);
               }
               
               @Override
               public CodeTemplate[] newArray(int ji) {
                    return new CodeTemplate[ji];
               }
          };
          
      public static final String PACKAGE_NAME = "${packageName}";
      public static final String CLASS_NAME = "${className}";
      
      protected String codeContent;
          
      public CodeTemplate() {}
          
      public CodeTemplate(Parcel in) {
           codeContent = in.readString();
      }
          
      public final String get() {
           configure();
           return codeContent;
      }
      
      public final void setContent(String contents) {
           codeContent = contents;
      }
      
      public void configure() {}
      
      public String getName() {
           throw new IllegalStateException("getName() is not subclassed");
      }
      
      public String getExtension() {
           throw new IllegalStateException("getExtension() is not subclassed");
      }
      
      @NonNull
      @Override
      public String toString() {
           return getName();
      }
      
      @Override
      public int describeContents() {
           return 0;
      }
      
      @Override
      public void writeToParcel(Parcel parcel, int ji) {
           parcel.writeString(codeContent);
      }
}

File: feature/feature-template/src/main/java/org/gampiot/robok/feature/template/code/CodeTemplate.kt
package org.gampiot.robok.feature.template.code

import android.os.Parcel
import android.os.Parcelable

open class CodeTemplate() : Parcelable {

    public var PACKAGE_NAME: String? = null
    public var CLASS_NAME: String? = null
    public var codeContent: String? = null

    constructor(parcel: Parcel) : this() {
        codeContent = parcel.readString()
    }

    override fun writeToParcel(parcel: Parcel, flags: Int) {
        parcel.writeString(codeContent)
    }

    override fun describeContents(): Int {
        return 0
    }

    open fun get(): String? {
        configure()
        return codeContent
    }

    open fun setContent(contents: String) {
        codeContent = contents
    }

    open fun configure() {}

    open fun getName(): String {
        throw IllegalStateException("getName() is not subclassed")
    }

    open fun getExtension(): String {
        throw IllegalStateException("getExtension() is not subclassed")
    }

    override fun toString(): String {
        return getName()
    }

    companion object CREATOR : Parcelable.Creator<CodeTemplate> {
        override fun createFromParcel(parcel: Parcel): CodeTemplate {
            return CodeTemplate(parcel)
        }

        override fun newArray(size: Int): Array<CodeTemplate?> {
            return arrayOfNulls(size)
        }
    }
}

